ENH: Use entry_points to install the f2py scripts.

charris opened the pull request:
This adds entry_points for the f2py scripts. The installed scripts
differ between Windows and other environments.

* On Windows, the only script installed is 'f2py'. This works well in
  that environment because each Python version is installed in its own
  directory, making it easy to keep the differing script versions
  separate.

* Otherwise, three scripts are installed, 'f2py', 'f2py' + 'minor', and
  'f2py' + 'major.minor'. For instance, if Numpy is installed by
  Python 2.7, then the installed scripts will be named 'f2py', 'f2py2',
  and 'f2py2.7'. That naming scheme is used for back compatibility, and
  also so that more than one Python version can be dealt with in a way
  common to many Linux distros. Note that 'f2py' will always point to
  the latest install and 'f2py(2|3)' to the latest Python (2|3) install

The script tests have been modified to check for the new environment
and the code previously used to install the scripts has been removed.

<!-- Please be sure you are following the instructions in the dev guidelines
http://www.numpy.org/devdocs/dev/gitwash/development_workflow.html
-->

<!-- We'd appreciate it if your commit message is properly formatted
http://www.numpy.org/devdocs/dev/gitwash/development_workflow.html#writing-the-commit-message
-->

charris committed:
ENH: Use entry_points to install the f2py scripts.

This adds entry_points for the f2py scripts. The installed scripts
differ between Windows and other environments.

* On Windows, the only script installed is 'f2py'. This works well in
  that environment because each Python version is installed in its own
  directory, making it easy to keep the differing script versions
  separate.

* Otherwise, three scripts are installed, 'f2py', 'f2py' + 'minor', and
  'f2py' + 'major.minor'. For instance, if Numpy is installed by
  Python 2.7, then the installed scripts will be named 'f2py', 'f2py2',
  and 'f2py2.7'. That naming scheme is used for back compatibility, and
  also so that more than one Python version can be dealt with in a way
  common to many Linux distros. Note that 'f2py' will always point to
  the latest install and 'f2py(2|3)' to the latest Python (2|3) install

The script tests have been modified to check for the new environment
and the code previously used to install the scripts has been removed.

eric-wieser commented on numpy/tests/test_scripts.py:
```diff
@@ -62,32 +62,35 @@ def run_command(cmd, check_code=True):
+        assert_(success == 1, msg)
     else:
+        # Three scripts are installed in unixlike systems.
+        # 'f2py', 'f2py.major', and 'f2py.major.minor'.
```
This comment is wrong - it's `f2py<major>` not `f2py.<major>`

eric-wieser commented on numpy/tests/test_scripts.py:
```diff
@@ -62,32 +62,35 @@ def run_command(cmd, check_code=True):
 def test_f2py():
     # test that we can run f2py script
+
+    def try_f2py_commands(cmds):
```
Out of scope for this change, but - this would make more sense as a `pytest.parametrize` rather than silencing the exceptions and counting.

charris replied:
Yep, I think the test and supporting function can probably use some cleanup at some point. I probably needs some work from someone running Windows to check if that part can also be improved.

eric-wieser commented:
This looks pretty good to me - although I think it's valuable to keep a `__main__.py` just so that I can invoke the f2py corresponding to a particular python interpreter with `python -m f2py`

eric-wieser commented on numpy/f2py/__main__.py:
```diff
@@ -1,27 +0,0 @@
-    sys.argv.append("-DNUMARRAY")
-    from f2py2e import main
-elif mode == "2e-numpy":
-    from numpy.f2py import main
```
What happened to these other modes?

charris commented:
Fixed for running the module. I am unable to figure out how to test for that.

eric-wieser commented:
Hasn't this removed support for `f2py --2e-numpy` and `f2py --2e-numeric`?

charris committed:
DOC: Note f2py script extension change on Windows.

Using console_scripts changes the type of the f2py script on Windows
from `*.py` to `*.exe`, so the script can no longer be executed as a
python script. This is a compatibility break and needs mentioning in the
release notes.

charris commented:
> Hasn't this removed support for f2py --2e-numpy and f2py --2e-numeric

Yes, but only so far as they are unrecognized options. Numarray is gone, and the same code compiles both numpy and numeric, although I have no idea if the latter still works. I'd like to get rid of all those options, but not in this PR. I suppose we could move the check for the options over to f2py2e, which is what ends up getting called regardless of the option.

charris commented:
Note that calling `__main__` in the console scripts causes a hang in some circumstances.

charris commented:
OK, I moved the argument checks. If that is acceptable I will squash the commit.

charris commented:
Ha, it turns out that none of those arguments have been working since whenever. Maybe we should just remove them?

charris commented:
I'm going to put this in so it gets some testing. @jsolbrig I've tested this locally and it doesn't hang, but it would be good to get some confirmation from folks who had the original problem. I'd add that test here, but it isn't well supported by the current f2py testing framework and would be more work than I want to undertake for this.

eric-wieser commented:
Seems that you merging caused my approval comment to be discarded - this looks pretty good to me anyway.

eric-wieser commented:
Can you reference what the original problem was?

jsolbrig commented:
@charris I just cloned the master branch and tested.  Everything seems to work when building my application with f2py.

charris commented:
@jsolbrig Great! Thanks for testing.
